### PR TITLE
fix canary chunk infer bug

### PR DESCRIPTION
**Pull request by stevehuang52**
# What does this PR do ?

- fix canary chunk infer rare bug, where the last segment might have mismatched shape with pre-allocated buffer.
- update audio_segment.py to allow duration is None
- add lhotse support for relative audio paths

### CODE REVIEW

**Comment by stevehuang52** [comment]
jenkins

**Comment by stevehuang52** [comment]
jenkins

**Comment by stevehuang52** [comment]
jenkins

**Comment by stevehuang52** [comment]
jenkins

File: nemo/collections/common/data/lhotse/nemo_adapters.py [hunk 1]
```diff
@@ -74,7 +75,7 @@ def path(self) -> str | Path:
 
     def __iter__(self) -> Generator[Cut, None, None]:
         for data in self.source:
-            audio_path = data.pop("audio_filepath")
+            audio_path = get_full_path(str(data.pop("audio_filepath")), str(self.path))
```
Comment (stevehuang52) [inline nemo/collections/common/data/lhotse/nemo_adapters.py:78]: @pzelasko I added the change here

**Comment by stevehuang52** [comment]
> could you add unit tests that cover these edge cases? btw I didn't see that PR earlier and added the relative audio path support independently; we can either merge my PR or port the test I wrote to yours and close mine (up to you) #8464

Sure I'll add the tests~I think it's better to use the `get_full_path` function to resolve the relative paths, since it's well tested.

File: nemo/collections/asr/parts/preprocessing/segment.py [hunk 2]
```diff
@@ -241,7 +241,7 @@ def from_file(
                     sample_rate = f.samplerate
                     if offset > 0:
                         f.seek(int(offset * sample_rate))
-                    if duration > 0:
+                    if duration is not None and duration > 0:
```
Comment (nithinraok) [inline nemo/collections/asr/parts/preprocessing/segment.py:244]: This is accepted change however I am curious if it did raise `None cannot be compared with integer` error? I use duration as null for all evaluations with speaker models and they run fine?

Comment (stevehuang52) [reply nemo/collections/asr/parts/preprocessing/segment.py:244]: hmm, interesting....at least with python3.10 it raised the error when I tried

**Review by nithinraok — APPROVED** [review]
LGTM.

**Comment by stevehuang52** [comment]
jenkins

**Comment by stevehuang52** [comment]
jenkins

**Comment by stevehuang52** [comment]
jenkins

**Review by pzelasko — APPROVED** [review]
LGTM

**Comment by stevehuang52** [comment]
jenkins
